runtime/runtime/src/lib.rs

```diff
@@ -948,29 +960,80 @@ impl Runtime {
         let mut validator_proposals = vec![];
         let mut local_receipts = vec![];
         let mut outcomes = vec![];
+        let mut total_gas_burnt = 0;
 
         for signed_transaction in transactions {
-            outcomes.push(self.process_transaction(
+            let outcome_with_id = self.process_transaction(
                 &mut state_update,
                 apply_state,
                 signed_transaction,
                 &mut local_receipts,
                 &mut new_receipts,
                 &mut stats,
-            )?);
+            )?;
+            total_gas_burnt += outcome_with_id.outcome.gas_burnt;
+
+            outcomes.push(outcome_with_id);
         }
 
-        for receipt in local_receipts.iter().chain(prev_receipts.iter()) {
+        let mut delayed_receipts_indices: DelayedReceiptIndices =
+            get(&state_update, DELAYED_RECEIPT_INDICES)?.unwrap_or_default();
+        let mut delayed_receipts_changed = false;
+
+        let mut process_receipt = |receipt: &Receipt,
+                                   state_update: &mut TrieUpdate,
+                                   total_gas_burnt: &mut Gas|
+         -> Result<_, StorageError> {
             self.process_receipt(
-                &mut state_update,
+                state_update,
                 apply_state,
                 receipt,
                 &mut new_receipts,
                 &mut validator_proposals,
                 &mut stats,
             )?
             .into_iter()
-            .for_each(|outcome_with_id| outcomes.push(outcome_with_id));
+            .for_each(|outcome_with_id| {
+                *total_gas_burnt += outcome_with_id.outcome.gas_burnt;
+                outcomes.push(outcome_with_id);
+            });
+            Ok(())
+        };
+
+        while delayed_receipts_indices.first_index < delayed_receipts_indices.next_available_index {
+            if total_gas_burnt >= apply_state.gas_limit {
```

Comment:
It seems like with this logic and `if total_gas_burnt < apply_state.gas_limit {` logic below `gas_limit` can be exceeded, (and will be exceeded in most cases) `test_apply_delayed_receipts` seems to demonstrate it too. Since `test_apply_delayed_receipts` assigns 1 gas_limit and we exceed it with every `apply` but still process one receipt on each step it takes exactly n steps to process all the receipts.

I think it is safer if our logic actually prevents gas_limit to be exceeded.

Reply:
Firstly, the Runtime might stuck with not receipt being produced due to large receipt that would exceed the gas_limit.

We can't estimate the gas_burnt for a receipt before processing it. If we already processed this receipt, there are no reason to not include it into the block.
Challenges would follow the same logic and always produce the same amount of receipts.